### PR TITLE
Bug Fixes

### LINKED ISSUE #10: remove line break in header of Parturition_output.csv
"Numbers_of detected_events" has a line break in the header.

### LINKED ISSUE #19: missing ids and blank rows in parturition_output.csv
We noticed three different types of missing information in the output table. I'll put an example of all of them here in case the causes are related:

1. Some parturition events are missing an Individual_id. It looks like for each Track_id, the corresponding Individual_id is present only for the first detected event:
image

2. In other cases, the Individual_id is present but the corresponding Track_id is missing:
image

3. And in other cases, the entire row is empty:
image

### LINKED ISSUE #22: Output not matching that of previous app
This probably relates to the switch to move2. It looks like the animal IDs and taxa are not being passed onto the next app.

In the app just before the Parturitions app, the cargo agent shows there are 10 animals, 3961 positions, taxa _Cervus elaphus_:  

image   

image   

image  

After the Parturitions app runs, the cargo agent shows 1 animal, 3937 positions, and no animal attributes, animal names, or taxa names available:

image   

image   

image  

In the output RDS, there is `individual.name.deployment.id` and `trackID` but no individual.local.identifier or animal.id. 

Might relate to [issue 19](https://github.com/nilanjanchatterjee/parturition/issues/19)?

### PR DESCRIPTION
**Pull request by smthfrmn**
This PR does some refactoring and bug fixes, namely:

- Removes duplicate code for when `threshold` is passed by user or uses default
- Adds some unit tests and sdk support
- Catches error in #25 and tells the user how to adjust the settings to fix it
- Fixes #22 
- Fixes #19 (I think...we need to verify this)
- Fixes #10

@nilanjanchatterjee I included comments throughout the code explaining what I did

### CODE REVIEW

File: RFunction.R [hunk 1]
```diff
@@ -7,279 +7,233 @@ library(magic)
...
-  class(data$distance) <-"numeric"
-  data_df <-as.data.frame(data)
-  names(data_df) <- gsub("[.]", "_", names(data_df))
-  # uid <-unique(data_df$individual_local_identifier)
-  uid <-unique(data$trackID) 
-  
-  ###  Function for plotting the individual speed
-  plot_speed <-function(dat, dat_outp, yul=yaxs_limit)
+###  Function for plotting the individual speed
+plot_speed <- function(dat, dat_outp, yul, track_id, threshold) {
```
Comment (smthfrmn) [inline RFunction.R:14]: Move plotting functions outside of main function to make main function more readable

File: RFunction.R [hunk 2]
```diff
@@ -7,279 +7,259 @@ library(magic)
...
-         ylab= expression(paste("Distance /", Delta, "t")), xlab= "Time", col= "grey40")
-    lines(dat$timestamp, dat$speed,col= "grey30", main= paste(uid[i], yr, sep = "_"))
-    lines(dat$timestamp, dat$rollm, col ="brown4", lwd=1.5, main= paste(uid[i], yr, sep = "_"))
-    #legend('topright', legend = rp, bty = 'n')
-    abline(h= ifelse(is.null(threshold), mean(dat$speed, na.rm=T), threshold),  lty=3, lwd=2, col= "coral")
-    for (i in 1:nrow(dat_outp))
-    {abline(v= dat_outp$V5, lty=2, lwd=1.5, col= "green4")
-      abline(v= dat_outp$V6, lty=4, lwd=1.5, col= "royalblue")
-    }
+
```
Comment (smthfrmn) [inline RFunction.R:10]: Move plotting functions outside of main function to make main function more readable

File: RFunction.R [hunk 3]
```diff
@@ -7,279 +7,259 @@ library(magic)
...
-      ### Input condition for the clustering 
-      ### used a mean criteria here can also be used a user specified cutoff 
-      data_temp$cnd <- ifelse((data_temp$speed) < mean(data_temp$rollm, na.rm=T) & !is.na(data_temp$speed),1,0)
-      
+}
+
+
+rFunction <- function(data, threshold = NULL, window = 72, yaxs_limit = 1000) {
+  original_track_id_column <- mt_track_id_column(data)
+  track_attribute_data <- mt_track_data(data)
```
Comment (smthfrmn) [inline RFunction.R:67]: Save track attribute data for returning object at the end

File: RFunction.R [hunk 4]
```diff
@@ -7,279 +7,259 @@ library(magic)
...
+    track_data <- data_df |>
+      filter(trackID == track_id)
+
+    tint <- as.numeric(
+      as.POSIXct(max(track_data$timestamp)) - as.POSIXct(min(track_data$timestamp)),
+      units = "hours"
+    )
+
+    if (nrow(track_data) > 10 & tint > window) { ## To filter individuals with very few relocations
+      data_temp <- tryCatch(
```
Comment (smthfrmn) [inline RFunction.R:112]: Doing all this in one dplyr function

File: RFunction.R [hunk 5]
```diff
@@ -7,279 +7,259 @@ library(magic)
...
+              window / median(as.numeric(timediff), na.rm = T),
+              fill = NA
+            ),
+            rollnsd = rollmean(nsd,
+              window / median(as.numeric(timediff), na.rm = T),
+              fill = NA
+            )
+          ),
+        error = function(e) {
+          if (grepl("rollmean", e$message)) {
```
Comment (smthfrmn) [inline RFunction.R:146]: Catch error that can happen if window is too small

File: RFunction.R [hunk 6]
```diff
@@ -7,279 +7,259 @@ library(magic)
...
+    paste("Parturition_vel", window, ".pdf")
+  ), width = 8, height = 12)
+
+  par(mfrow = c(4, 3), mar = c(4, 4, 3, 1))
+
+
+  for (i in 1:length(track_ids)) {
+    track_id <- track_ids[i]
+
+    animal_id <- track_attribute_data |>
```
Comment (smthfrmn) [inline RFunction.R:96]: getting animal id

File: RFunction.R [hunk 7]
```diff
@@ -7,279 +7,259 @@ library(magic)
...
+    drop_na(Start_date)
+
+  # write app artefact
+  write.csv(dat_final_output, file = paste0(
+    app_artifacts_base_path,
+    paste("Parturition_output", window, ".csv")
+  ))
+
+  # convert the data.frame output into move2 object
+  dat_final <- left_join(
```
Comment (smthfrmn) [inline RFunction.R:246]: join track attribute data back with data from function

File: tests/testthat/test_RFunction.R [hunk 8]
```diff
@@ -0,0 +1,39 @@
+source(here("tests/testthat/helper.R"))
```
Comment (smthfrmn) [inline tests/testthat/test_RFunction.R:1]: some unit tests to make sure function executes

**Review by nilanjanchatterjee — APPROVED** [review]

**Comment by nilanjanchatterjee** [comment]
Gone through the changes you have made and most of them make sense to me...I'll do some more through testing probably with some edge cases...Great work @smthfrmn
